### DISCUSSION TITLE
Initial setup for very slow system? (Underfloor heating in old brick building)

### DISCUSSION THREAD

**Discussion by freaked1234** [ScratMan/HASmartThermostat#112] [Q&A]
Hey there!

I am living in a massive brick building with underfloor (water) heating that is very slow to react. Could you give me some guidance how to initially setup this thermostat and how to fine tune it over time?

Thank you very much!

**Comment by ScratMan** [comment]
Hello, if you already have temperature measurements on which you can see the oscillation due to regulation, you may compute the PID gains using the method described here :
https://github.com/ScratMan/HASmartThermostat/discussions/21#discussioncomment-1709437

**Comment by 10bn** [comment]
@freaked1234 Did figure this out for yourself and could you may share your settings?

**Comment by freaked1234** [comment]
> @freaked1234 Did figure this out for yourself and could you may share your settings?

Hey, since winter was pretty much over when I wrote this I don't know if my settings work. I'll play around a bit with it when it is getting colder and will let you know.

Maybe we could exchange contact details and work something out together. First we should compare our building specs to see how much difference there is. A throwaway mail of mine is aerielpally@gmail.com

Anyway I'll post my settings as soon as I know they work and I would appreciate if you could do the same!

**Comment by T81** [comment]
Some useful info would be the temperature of inlet water (if known) and if it is easy to overshoot the setpoint temperature

**Comment by T81** [comment]
@freaked1234 any updates?

**Comment by HermesHonshappo** [comment]
Here are the values I have for a room with underfloor heating, using water.
- Kp: 40
- Ki:  0.00075
- Kd: 30000
- Ke: 1.5

Room is insulated basically: sheetrock with 8cm insulation behind, and when the floor was poured, there is 8cm insulation below the floor with the water pipes.  But the walls behind the insulation are basic brick, the windows are basic double glass windows (but the cheapest ones in the store), and the ceiling has no insulation yet

The water inlet is regulated to 35°C (I have a mechanical temp controlled mixer & pump purchased from Amazon)

I get regulation I'd say is acceptable:

<img width="393" alt="Screenshot 2023-01-11 at 14 17 12" src="https://user-images.githubusercontent.com/97752752/211815914-c078f2cb-3597-42f3-ad75-c3aad636c586.png">

EDIT: one thing I forgot to add: I have a temp sensor that is in the floor slab. And I'm regulating not by measuring temp in the room, but temp of the floor slab. I found that having the floor slab at 25°C during the day gives an acceptable room temp (between 18 and 19°C), and good comfort

**Reply by cctl01** [reply]
> EDIT: one thing I forgot to add: I have a temp sensor that is in the floor slab. And I'm regulating not by measuring temp in the room, but temp of the floor slab. I found that having the floor slab at 25°C during the day gives an acceptable room temp (between 18 and 19°C), and good comfort 

This comment opened up a new can of worms!

**Reply by tomandegg** [reply]
Thanks for sharing. What values do you use for pwm and min_cycle_duration?

**Reply by HermesHonshappo** [reply]
I can't remember what I had at that time. In my old savefiles, it looks like I used 30 mins for PWM, 3 mins min cycle and 5 minute sampling period

Since then, I have decided to experiment with using the temp sensor in the room and regulating directly with the room temperature
Here are the values I use now:
```
- platform: smart_thermostat
  name: Thermostat Ch. Ian
  unique_id: 7d503af0-0100-4512-995f-1cea59d7e636
  heater: input_boolean.commande_pompe_1
  target_sensor: sensor.temperature_ch_ian_temperature
  target_temp_step: 0.1
  ac_mode: False
  keep_alive:
    seconds: 60
  pwm: 00:30:00
  sampling_period: 00:15:00
  min_cycle_duration: 00:03:00
  cold_tolerance: 0.1
  hot_tolerance: 0.1
  min_temp: 10
  max_temp: 24
  away_temp: 12
  sleep_temp: 16
  comfort_temp: 18
  preset_sync_mode: sync
  kp : 30
  ki : 0.0008
  kd : 25000
  outdoor_sensor: sensor.temperature_exterieur_temperature
  ke: 2
```

**Reply by tomandegg** [reply]
Thanks for posting your config. Looking at your graph above and relating it with your config's pwm and min_cycle_duration, I think I am talking about another level of slow reacting heating system in my place. It's a very, very well insulated house and it does not get really cold even if I turn off the heating completely and the water in the floor heating is also not that warm, at least the maximum floor temperature I ever measured was around 28 or 29 degrees in deep winter.

Which means, as soon as the floor temperature overshoots the set temperature, the rooms get much too hot and take a lot of time to cool back down. Another hurdle is that my valves react a lot slower than yours, opening and closing each take almost 5 minutes if the red bar that peaks out and retracts very slowly means anything.

So what I need is p/i/d such that the actual temperature approaches the set temperature very slowly. But I assume your values might be a good starting point to see how the system behaves and to further tweak the values then according to theory.

**Reply by HermesHonshappo** [reply]
When I was using regulation based on the floor temp sensor, I very very rarely had overshoot of the temperature. I think it happened a few times, and some were my mistakes.

What bothered me is that in late winter/early spring, the outside temp rose, and then a floor temp or 25 resulted in a room temp too high.

Now that I use room temp, I do have overshoot, but it's related to sun exposure. When there's a bright sun shining through the window, the room temp goes quickly high above the target of 19°C. But on the bright side, since the overshoot is detected early, the heating drops to 0% output quickly, and floor temp stops rising

And when the sun goes past the above mountain summit, the external temp drops rather quickly, the internal temp drops, the E and D components kick in, and the floor starts heating back before the room temp has dropped below the 19°C target

For example, this is today's chart:

![image](https://github.com/ScratMan/HASmartThermostat/assets/97752752/1b2e0193-8799-45f3-81f9-97245886a1a3)
red line is room temp, gray line floor temp, and light blue inlet water temp (when it leaves the pump towards the floor circuit)

_Ideally, I would have to write a formula that sets floor target temp taking into account the outside average temp.
I already have something similar, which I use to set target temps of other room thermostats.  When outside temp 48h avg is above 15, target comfort temp is 19°C, and if outside avg temp drops, I slowly lower the inside comfort temp. It also takes into account the kWh price, which is variable in my case.
In a nutshell, high kWh price and/or lower 48h outside temp avg-> lower target sleep & comfort presets is what I have currently implemented for all other rooms, where heating is conventional electric heaters._

_I'd have to write a reverse formula, like if outside avg temp is 5°, target floor temp should be 25, and if outside avg temp rises, I lower target floor temp, until they intersect.  I'm suspecting this would result in a near perfect regulation, year-round. But I've been busy automating other things in the house this year, so I haven't looked too deep into this._

**Back to your use case:**  you could increase the min cycle duration, but in that case I'd suggest increasing the PWM also. If not, when the target P output goes below 10% for example, your valves would open&close too often, sometimes starting a "close" while the previous "open" isn't finished.
The way I sized it is that my min cycle duration is dictated by the valves, and pwm should large enough to that min cycle would be no more than 10% of the pwm window.
So for a 7 min min cycle duration, (which I assume would be OK for a 5 min valve), you'd end up with a pwm window of 70 mins

**Reply by tomandegg** [reply]
My reasoning was similar and I went with a 7 min min cycle duration but I left the pwm at 30 min as I'd rather have more short pulses in order for the floor to keep warm. 

This works quite well so far, however, I'll probably have to wait for the next winter to evaluate in depth with colder outside temperatures.

Thanks for your suggestions, very much appreciated!
